src/Microsoft.ML.AutoML/Experiment/Experiment.cs

```diff
@@ -196,6 +196,22 @@ public IList<TRunDetail> Execute()
                         "was reached, and the running MLContext was stopped. Details: {0}", e.Message);
                     return iterationResults;
                 }
+                catch (AggregateException e)
+                {
+                    // This exception is thrown when the IHost/MLContext of the trainer is canceled due to
+                    // reaching maximum experiment time. Simply catch this exception and return finished
+                    // iteration results. For some trainers, like FastTree, because training is done in parallel
+                    // in can throw multiple OperationCancelledExceptions. This causes them to be returned as an
+                    // AggregateException and misses the first catch block. This is to handle that case.
+                    if (e.InnerExceptions.All( exception => exception is OperationCanceledException))
+                    {
+                        _logger.Warning("OperationCanceledException has been caught after maximum experiment time" +
+                        "was reached, and the running MLContext was stopped. Details: {0}", e.Message);
+                        return iterationResults;
+                    }
+
+                    throw;
+                }
             } while (_history.Count < _experimentSettings.MaxModels &&
                     !_experimentSettings.CancellationToken.IsCancellationRequested &&
                     !_experimentTimerExpired);
```

Comment:
### Method of passing log messages
In a related (non-filed) issue, to support the the ability to cancel runs, we needed a separate context for the child runs, and we pass the log messages from the child to parent. 

Current method of passing the messages is imperfect.

**Current returned log format:**
`"[Source=AutoML, Kind=Trace] [Source=ImageClassificationTrainer; ImageClassificationTrainer, Kind=Trace] Phase: Bottleneck Computation, Dataset used:      Train, Image Index: 3954"`

This duplicates the header on the messages (`[Source=AutoML, Kind=Trace]` and `[Source=ImageClassificationTrainer; ImageClassificationTrainer, Kind=Trace]`),  and makes it difficult to filter/select to the wanted log lines.

**Correct format:**
`"[Source=AutoML; ImageClassificationTrainer; ImageClassificationTrainer, Kind=Trace] Phase: Bottleneck Computation, Dataset used:      Train, Image Index: 3954"`

This single header then lets you filter on the log line properties (e.g. `e.Source == "xyz"`) instead of its free text.

Reply:
I logged this as a separate issue, since I'm not sure it'll be fixed while you're working on similar code here.

https://github.com/dotnet/machinelearning/issues/5632

Reply:
This doesn't seem like something I can change in this location though is it?